shape.c

```diff
@@ -499,13 +499,27 @@ get_next_shape_internal(rb_shape_t * shape, ID id, enum shape_type shape_type, b
 
     *variation_created = false;
 
+    // Fast path: if the shape has a single child, we can check it without a lock
+    struct rb_id_table *edges = RUBY_ATOMIC_PTR_LOAD(shape->edges);
+    if (edges && SINGLE_CHILD_P(edges)) {
```

Comment:
I think in the case where `edges` is `NULL` we could also allocate the child atomically, and set it with CAS.

Not sure if it's really worth it though, because supposedly code running in ractors produce stable shapes, so write would become rare quickly.

Reply:
I agree we could do that, an issue I think that would create though is that if two Ractors are trying to insert the exact same shape, we would end up creating two shapes for it, for one Ractor the CAS would fail (as is normal with this pattern), but then there'd be no ability to free the shape it had created, since they are never freed (though I suppose we could create a freelist specifically for this case).

Reply:
> (though I suppose we could create a freelist specifically for this case).

Yes, or just accept such case is rare enough that we can leak a few shapes.